examples/nlp/ner_transformers.py

```diff
@@ -0,0 +1,326 @@
+    all_labels = [(label1, label2) for label2 in ner_labels for label1 in iob_labels]
+    all_labels = ["-".join([a, b]) for a, b in all_labels]
+    all_labels = ["[PAD]", "O"] + all_labels
+    return dict(zip(range(0, len(all_labels) + 1), all_labels))
+
+
+mapping = make_tag_lookup_table()
+mapping
+
+"""
+Get a list of all tokens in the training dataset. This will be used to create the
+vocabulary
+"""
+
+all_tokens = sum(conll_data["train"]["tokens"], [])
+all_tokens_array = np.array(list(map(str.lower, all_tokens)))
+
+counter = Counter(all_tokens_array)
+len(counter)
+
+num_tags = len(mapping)
+vocab_size = 20000
+vocabulary = [token for token, count in counter.most_common(vocab_size - 2)]
+
+# The StringLook class will convert tokens to token IDs
+lookup_layer = tf.keras.layers.experimental.preprocessing.StringLookup(
+    vocabulary=vocabulary
+)
+
+"""
+Create 2 new `Dataset` objects from the training and validation data
+"""
+
+train_data = tf.data.TextLineDataset("./data/conll_train.txt")
+val_data = tf.data.TextLineDataset("./data/conll_val.txt")
+
+# Print out one line to make sure it looks good
+# The first record in the line is the number of tokens. After that we will
+# have all the tokens followed by all the ner tags
+
+list(train_data.take(1).as_numpy_iterator())
+
+"""
+We will be using the following map function to transform the data in the dataset:
+"""
+
+
+def map_func(record):
+    record = tf.strings.split(record, sep="\t")
+    length = tf.strings.to_number(record[0], out_type=tf.int32)
+    tokens = record[1 : length + 1]
+    tokens = tf.strings.lower(tokens)
+    tags = record[length + 1 :]
+    tags = tf.strings.to_number(tags, out_type=tf.int64)
+    tags += 1
+    return tokens, tags
+
+
+batch_size = 32
+train_dataset = (
+    train_data.map(map_func)
+    .map(lambda x, y: (lookup_layer(x), y))
+    .padded_batch(batch_size)
+)
+val_dataset = (
+    val_data.map(map_func)
+    .map(lambda x, y: (lookup_layer(x), y))
+    .padded_batch(batch_size)
+)
+
+ner_model = NERModel(num_tags, vocab_size, embed_dim=32, num_heads=4, ff_dim=64)
+
+"""
+We will be using a custom loss function that will ignore the loss from padded tokens.
+"""
+
+
+class CustomLoss(tf.keras.losses.Loss):
+    def __init__(self, name="custom_ner_loss"):
+        super().__init__(name=name)
+
+    def call(self, y_true, y_pred):
+        loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(
+            from_logits=True, reduction=tf.keras.losses.Reduction.NONE
+        )
+        loss = loss_fn(y_true, y_pred)
+        mask = tf.cast((y_true > 0), dtype=tf.float32)
+        loss = loss * mask
+        return tf.reduce_sum(loss) / tf.reduce_sum(mask)
+
+
+loss = CustomLoss()
+
+"""
+## Compile and fit the model
+"""
+
+ner_model.compile(optimizer="adam", loss=loss)
+
+ner_model.fit(train_dataset, epochs=10)
+
+sample_input = lookup_layer(
+    "eu rejects german call to boycott british lamb".lower().split()
```

Comment:
Consider moving `.lower().split()` to a separate tokenization function to make it reusable (and also include the lookup op as part of it)